File: langs/typescript/src/commands/modeledCrudCommands.ts
/**
 * ModeledCrudCommands — Create / UpdateCurrent / DeleteCurrent helper.
 *
 * See spec/06-composite-vm.md §Modeled CRUD commands and ADR-0016.
 */
import {
  ConfirmationDecoratorCommand,
  type ConfirmDelegate,
} from "./confirmationDecoratorCommand.js";
import { RelayCommand } from "./relayCommand.js";
import type { ICommand } from "./types.js";

// The `M` (model) type parameter is a phantom: the helper only manipulates
// VMs at runtime, but exposing `M` at the type boundary keeps the public
// signature symmetrical with C# `ModeledCrudCommands<M, VM>` and Python
// `ModeledCrudCommands[M, VM]` per ADR-0006 / ADR-0016.
export interface ModeledCrudCommandsOptions<M, VM> {
  current: () => VM | null | undefined;
  createNew: () => void;
  updateCurrent: (vm: VM) => void;
  deleteCurrent: (vm: VM) => void;
  confirmUpdate?: ConfirmDelegate;
  confirmDelete?: ConfirmDelegate;
  /**
   * @internal Phantom field to pin the `M` type parameter so it is not erased
   * by the TypeScript compiler. Never read at runtime; always undefined.
   */
  _model?: M;
}

export class ModeledCrudCommands<M, VM> {
  readonly createNewCommand: ICommand;
  readonly updateCurrentCommand: ICommand;
  readonly deleteCurrentCommand: ICommand;

  // Inner RelayCommands hold trigger subscriptions; track them so dispose()
  // can tear them down (parity with C# ModeledCrudCommands.Dispose).
  readonly #innerRelays: readonly RelayCommand[];
  #disposed = false;

  constructor(opts: ModeledCrudCommandsOptions<M, VM>) {
    const create = RelayCommand.builder().task(opts.createNew).build();
    const update = RelayCommand.builder()
      .task(() => {
        const c = opts.current();
        if (c) opts.updateCurrent(c);
      })
      .predicate(() => opts.current() != null)
      .build();
    const remove = RelayCommand.builder()
      .task(() => {
        const c = opts.current();
        if (c) opts.deleteCurrent(c);
      })
      .predicate(() => opts.current() != null)
      .build();

    this.#innerRelays = [create, update, remove];

    this.createNewCommand = create;
    this.updateCurrentCommand = opts.confirmUpdate
      ? new ConfirmationDecoratorCommand(update, opts.confirmUpdate)
      : update;
    this.deleteCurrentCommand = opts.confirmDelete
      ? new ConfirmationDecoratorCommand(remove, opts.confirmDelete)
      : remove;
  }

  /**
   * Dispose the inner RelayCommands and any confirmation wrappers. Idempotent.
   *
   * The public update/delete commands may be `ConfirmationDecoratorCommand`
   * wrappers (when `confirmUpdate` / `confirmDelete` are supplied). Each wrapper
   * owns an `errors` Subject whose contract is to complete on dispose, so the
   * wrappers are disposed here alongside the inner relays (parity with
   * C#/Python). When no confirm hook is supplied the public command is the inner
   * relay itself, already covered.
   */
  dispose(): void {
    if (this.#disposed) return;
    this.#disposed = true;
    const targets: { dispose(): void }[] = [...this.#innerRelays];
    for (const cmd of [this.updateCurrentCommand, this.deleteCurrentCommand]) {
      if (cmd instanceof ConfirmationDecoratorCommand) targets.push(cmd);
    }
    for (const cmd of targets) cmd.dispose();
  }
}

File: langs/swift/Sources/VMx/Commands/ModeledCrudCommands.swift
//
// ModeledCrudCommands — Create / UpdateCurrent / DeleteCurrent helper.
//
// See spec/06-composite-vm.md §7 and ADR-0016.
//
// Note: Swift drops the phantom `M` type parameter present in C# / Python / TS
// (see ADR-0006 §3.2 — phantom types are not idiomatic in Swift). The helper
// is parameterised only on VM (the view-model type).
//
import Foundation
import Combine

/// Packages the three canonical CRUD commands — `createNewCommand`,
/// `updateCurrentCommand`, `deleteCurrentCommand` — over a "currently
/// selected" view-model of type `VM`.
///
/// - `createNewCommand`: always executable; runs `createNew`.
/// - `updateCurrentCommand`: enabled only when `current()` returns non-nil;
///   runs `updateCurrent(vm)`. Optionally gated by a `confirmUpdate` delegate.
/// - `deleteCurrentCommand`: enabled only when `current()` returns non-nil;
///   runs `deleteCurrent(vm)`. Optionally gated by a `confirmDelete` delegate.
///
/// Pass `currentChanged` to wire a trigger so `canExecuteChanged` fires
/// whenever the selection changes and consumers can refresh button state.
public final class ModeledCrudCommands<VM: AnyObject> {
    public let createNewCommand: Command
    public let updateCurrentCommand: Command
    public let deleteCurrentCommand: Command

    // Inner RelayCommands hold trigger subscriptions; tracked so dispose() can
    // tear them down (parity with C# ModeledCrudCommands.Dispose).
    private let innerRelays: [RelayCommand]
    private var disposed = false

    public init(
        current: @escaping () -> VM?,
        createNew: @escaping () -> Void,
        updateCurrent: @escaping (VM) -> Void,
        deleteCurrent: @escaping (VM) -> Void,
        confirmUpdate: (() async throws -> Bool)? = nil,
        confirmDelete: (() async throws -> Bool)? = nil,
        currentChanged: AnyPublisher<Void, Never>? = nil
    ) {
        // create: no predicate — always executable.
        let create = RelayCommand.builder()
            .task(createNew)
            .build()

        // update: gated on current() != nil; optional currentChanged trigger.
        var updateBuilder = RelayCommand.builder()
            .task({ if let c = current() { updateCurrent(c) } })
            .predicate({ current() != nil })
        if let changed = currentChanged {
            updateBuilder = updateBuilder.triggers(changed)
        }
        let update = updateBuilder.build()

        // delete: gated on current() != nil; optional currentChanged trigger.
        var deleteBuilder = RelayCommand.builder()
            .task({ if let c = current() { deleteCurrent(c) } })
            .predicate({ current() != nil })
        if let changed = currentChanged {
            deleteBuilder = deleteBuilder.triggers(changed)
        }
        let remove = deleteBuilder.build()

        innerRelays = [create, update, remove]

        createNewCommand = create
        updateCurrentCommand = confirmUpdate.map { gate in
            ConfirmationDecoratorCommand(update, confirm: gate)
        } ?? update
        deleteCurrentCommand = confirmDelete.map { gate in
            ConfirmationDecoratorCommand(remove, confirm: gate)
        } ?? remove
    }

    /// Dispose the underlying RelayCommands and their trigger subscriptions.
    /// Idempotent: subsequent calls are a no-op.
    ///
    /// Note: `ConfirmationDecoratorCommand` wrappers (when confirm delegates are
    /// supplied) are NOT tracked separately because they hold no subscriptions of
    /// their own — `canExecuteChanged` is a direct passthrough to the inner
    /// RelayCommand's `canExecuteChanged`. The RelayCommand's dispose() tears
    /// down the trigger subscriptions.
    public func dispose() {
        guard !disposed else { return }
        disposed = true
        for cmd in innerRelays { cmd.dispose() }
    }
}
